File: base/src/main/java/com/zhangteng/base/base/BaseApplication.java
package com.zhangteng.base.base;

import android.app.Activity;
import android.app.Application;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created by swing on 2018/8/2.
 * 应用启动时需要初始化的操作可继承BaseApplication实现相关方法
 */
public abstract class BaseApplication extends Application {
    public static BaseApplication instance;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
    }

    public static BaseApplication getInstance() {
        return instance;
    }

    /**
     * Module application初始化
     */
    public abstract void initModuleApp(Application application);

    /**
     * Module 数据初始化
     */
    public abstract void initModuleData(Application application);

    /**
     * 获取正在运行的activity
     */
    public static List<Activity> getActivitiesByApplication() {
        List<Activity> list = new ArrayList<>();
        try {
            Class<Application> applicationClass = Application.class;
            Field mLoadedApkField = applicationClass.getDeclaredField("mLoadedApk");
            mLoadedApkField.setAccessible(true);
            Object mLoadedApk = mLoadedApkField.get(getInstance());
            Class<?> mLoadedApkClass = mLoadedApk.getClass();
            Field mActivityThreadField = mLoadedApkClass.getDeclaredField("mActivityThread");
            mActivityThreadField.setAccessible(true);
            Object mActivityThread = mActivityThreadField.get(mLoadedApk);
            Class<?> mActivityThreadClass = mActivityThread.getClass();
            Field mActivitiesField = mActivityThreadClass.getDeclaredField("mActivities");
            mActivitiesField.setAccessible(true);
            Object mActivities = mActivitiesField.get(mActivityThread);
            // 注意这里一定写成Map，低版本这里用的是HashMap，高版本用的是ArrayMap
            if (mActivities instanceof Map) {
                @SuppressWarnings("unchecked")
                Map<Object, Object> arrayMap = (Map<Object, Object>) mActivities;
                for (Map.Entry<Object, Object> entry : arrayMap.entrySet()) {
                    Object value = entry.getValue();
                    Class<?> activityClientRecordClass = value.getClass();
                    Field activityField = activityClientRecordClass.getDeclaredField("activity");
                    activityField.setAccessible(true);
                    Object o = activityField.get(value);
                    list.add((Activity) o);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            list = null;
        }
        return list;
    }
}

File: base/src/main/java/com/zhangteng/base/base/BaseApplication.kt
package com.zhangteng.base.base

import android.app.Activity
import android.app.Application
import com.zhangteng.base.utils.isDebug
import java.util.*

/**
 * Created by swing on 2018/8/2.
 * 应用启动时需要初始化的操作可继承BaseApplication实现相关方法
 */
abstract class BaseApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        instance = this
    }

    /**
     * Module application初始化
     */
    abstract fun initModuleApp(application: Application?)

    /**
     * Module 数据初始化
     */
    abstract fun initModuleData(application: Application?)

    companion object {
        var instance: BaseApplication? = null

        /**
         * 获取正在运行的activity
         */
        fun getActivitiesByApplication(): MutableList<Activity?>? {
            var list: MutableList<Activity?>? = ArrayList()
            try {
                val applicationClass: Class<Application> = Application::class.java
                val mLoadedApkField = applicationClass.getDeclaredField("mLoadedApk")
                mLoadedApkField.isAccessible = true
                val mLoadedApk = mLoadedApkField[instance]
                val mLoadedApkClass: Class<*> = mLoadedApk.javaClass
                val mActivityThreadField = mLoadedApkClass.getDeclaredField("mActivityThread")
                mActivityThreadField.isAccessible = true
                val mActivityThread = mActivityThreadField[mLoadedApk]
                val mActivityThreadClass: Class<*> = mActivityThread.javaClass
                val mActivitiesField = mActivityThreadClass.getDeclaredField("mActivities")
                mActivitiesField.isAccessible = true
                val mActivities = mActivitiesField[mActivityThread]
                // 注意这里一定写成Map，低版本这里用的是HashMap，高版本用的是ArrayMap
                if (mActivities is MutableMap<*, *>) {
                    for ((_, value) in mActivities) {
                        val activityClientRecordClass: Class<*> = value!!::class.java
                        val activityField = activityClientRecordClass.getDeclaredField("activity")
                        activityField.isAccessible = true
                        val o = activityField[value]
                        list?.add(o as Activity)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                list = null
            }
            return list
        }
    }
}